# response error 처리하기

1. user-agent 수정하기
2. referer 수정하기

---
보통 1.user-agent를 수정하면 대부분 해결이 된다.

## 라이브러리 및 에러 페이지 준비하기

In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
import requests

%config InlineBackend.figure_formats={'png','retina'}
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
url = f'https://finance.daum.net/api/exchanges/summaries'

In [4]:
response = requests.get(url)
response

<Response [403]>

In [6]:
response.text

'<html>\r\n<head><title>403 Forbidden</title></head>\r\n<body bgcolor="white">\r\n<center><h1>403 Forbidden</h1></center>\r\n<hr><center>nginx</center>\r\n</body>\r\n</html>\r\n'

### user-agent 수정하기

여전히 403 에러 코드난 것을 확인할 수 있다.
보통의 경우 user-agent만 수정해도 해결 된다.

In [9]:
headers = {'User-Agent':
           'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36'}

response = requests.get(url, headers=headers)
response

<Response [403]>

### referer 수정한다

In [10]:
headers = {'User-Agent':
           'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36',
            'Referer':
            'https://finance.daum.net/exchanges'}

response = requests.get(url, headers=headers)
response

<Response [200]>

In [12]:
data = response.json()['data']
data

[{'symbolCode': 'FRX.KRWUSD',
  'date': '2023-08-24 14:42:06',
  'currencyCode': 'USD',
  'currencyName': '달러',
  'currencyUnit': 1,
  'country': '미국',
  'region': {'korName': '아메리카', 'engName': 'America'},
  'name': '미국 (USD/KRW)',
  'recurrenceCount': 484,
  'basePrice': 1321.5,
  'change': 'FALL',
  'changePrice': 11.5,
  'changeRate': 0.0086271568,
  'cashBuyingPrice': 1344.62,
  'cashSellingPrice': 1298.38,
  'ttBuyingPrice': 1308.6,
  'ttSellingPrice': 1334.4,
  'tcBuyingPrice': None,
  'fcSellingPrice': None,
  'exchangeCommission': 7.165,
  'usDollarRate': 1.0,
  'chartImageUrl': {'day': 'https://t1.daumcdn.net/finance/chart/kr/daumforex/d/FRX.KRWUSD.png',
   'month': 'https://t1.daumcdn.net/finance/chart/kr/daumforex/m/FRX.KRWUSD.png',
   'month3': 'https://t1.daumcdn.net/finance/chart/kr/daumforex/m3/FRX.KRWUSD.png',
   'year': 'https://t1.daumcdn.net/finance/chart/kr/daumforex/y/FRX.KRWUSD.png'}},
 {'symbolCode': 'FRX.KRWJPY',
  'date': '2023-08-24 14:42:06',
  'currencyCode

In [14]:
df = pd.DataFrame(data)[['symbolCode', 'currencyCode','country','currencyName','basePrice']]
df.tail(2)

symbolCode currencyCode country currencyName  basePrice
40  FRX.KRWQAR          QAR     카타르           리얄     362.59
41  FRX.KRWTRY          TRY      터키           리라      48.55